### ISSUE TITLE
ThemeBuilder command-line interface (CLI)

### ISSUE THREAD

**Issue by BingoRUS** [DevExpress/DevExtreme#5380]
# The Problem
* You cannot generate a new color scheme (CSS files) from theme metadata (theme color and other variables) programmatically or at build time.

* You cannot export a custom color scheme as a list of LESS or Sass variables that can be used in custom stylesheets.
# The Proposed Solution
We are introducing a set of DevExtreme CLI tools for generating themes.

### Build a custom color scheme
```bash 
> npx devextreme-cli build-theme [--base-theme][--input-file][--make-swatch][--output-file][--output-color-scheme][--help]
```
### Export theme variables as a less/scss file
```bash
> npx devextreme-cli export-theme-vars [--base-theme][--input-file][--output-format][--output-file][--base][--help]
```
### Export theme variables as a json metadata file
```bash
> npx devextreme-cli export-theme-meta [--base-theme][--input-file][--output-file][--base][--help]
```

### Installation

```bash
> npm install devextreme-cli -g
```

or

```bash
> yarn global add devextreme-cli
```

### Build a custom color scheme
```bash 
> devextreme build-theme [--base-theme][--input-file][--make-swatch][--output-file][--output-color-scheme][--help]
```
### Export theme variables as a less/scss file
```bash
> devextreme export-theme-vars [--base-theme][--input-file][--output-format][--output-file][--base][--help]
```
### Export theme variables as a json metadata file
```bash
> devextreme export-theme-meta [--base-theme][--input-file][--output-file][--base][--help]
```

## Command line arguments

* `--base-theme` – Specifies a base theme’s name (the default value is `generic.light`) 

* `--input-file` – Specifies the name of the file that contains input data (a .json file with metadata or a .less/.scss file with Bootstrap variables)

* `--output-file` – Specifies the output file name

* `--output-format` – Specifies the output file format (‘less’ or ‘scss’) (the default value is ‘less’ or the extension extracted from the `--output-file` value (if it contains any))

* `--output-color-scheme` – Specifies the custom color scheme name (the default value is `custom-scheme`)

* `--make-swatch` – If present, adds a CSS scope to each CSS rule (.dx-swatch-xxx), where xxx is the value from the `--output-color-scheme` parameter

* `--base` – Exports only base variables which are used to produce the derived variables

* `--help` – Displays help

## Examples

The following command generates a new `custom` color swatch using Generic Dark as a base theme:
```bash
> npx devextreme-cli build-theme –base-theme="generic.dark" --make-swatch --output-color-scheme="custom"
```

The following command generates a new `pink` color scheme based on Material Blue Light theme using constants from pink.json file:
```bash
> npx devextreme-cli build-theme --base-theme="material.blue.light" --input-file="pink.json" --output-color-scheme="pink"
```

The following command exports base variables for Generic Carmine theme in LESS format:
```bash
> npx devextreme-cli export-theme-vars --base-theme="generic.carmine" --output-format="less" --base
```

# We Need Your Feedback
## Take a Quick Poll
[Do the new CLI tools meet your needs?](https://docs.google.com/forms/d/e/1FAIpQLSeSv5o3Ig-D2O11S4eXT3mZtL72sqF6n0iazDEBrVFVtCg6Hw/viewform)

# Get Notified of Updates
Subscribe to this thread or our [Facebook](https://www.facebook.com/DevExpress.DevExtreme/) and [Twitter](https://twitter.com/devextreme) accounts for updates on this topic.

**Comment by lifeinchords** [comment]
Thanks for prioritizing this issue - it will help development significantly. 

I'm not sure I understand:

-  is the CLI a one-off command, generates SCSS files, which we then copy to our app scss directory and start modifying, in essence a fork of the DevEx SCSS? 

- Will the SCSS updates be tied to the DevEx node module? What happens when the official theme SCSS moves forward, say bugs are fixed and new components adde - do we need to re-export? how do we do an upstream merge?

- Any thoughts how the CLI fits into a build system workflow? We're using Webpack 4 now, but general workflow ideas will be helpful. We currently have 2 configs, one for production which compiles our app/custom css, with all 3rd party node module SCSS into a single scss bundle file, and a dev server, which listens for source changes and recompiles to memory.  

A barebones working example would be really helpful. Something like https://github.com/telerik/kendo-ui-npm-example

**Comment by BingoRUS** [comment]
Hi @lifeinchords and thank you for your interest!

As you can see in our [source code](https://github.com/DevExpress/DevExtreme/blob/18_2/styles/), the DevExtreme styles are organized as LESS files, where we have special ones to [describe different color schemes](https://github.com/DevExpress/DevExtreme/tree/18_2/styles/widgets/generic/color-schemes). We don't distribute these raw files, only the resulting CSS files. ThemeBuilder commands in DevExtreme CLI, as well as our online ThemeBuilder tool, give you the capability to:
- build a custom CSS file by replacing some of color scheme constants (for example, background color);
- export the same constants we use for our widgets to customize other (non-DevExtreme) parts of your application in LESS or Sass format.

So, answering your questions:
- If you use `export-theme-vars` command, the resulting SCSS file can be used **to customize your application**, not DevExtreme widgets. For example, you will be able to color all pure text on your page so it has the same color as the text inside a DevExtreme DataGrid.
- You can export all variables as metadata using the `export-theme-meta` command, modify them and use as an input file for the `build-theme` command. You will get a CSS file with a custom color scheme that can be used in your application **to customize DevExtreme widgets**.
- You need to build a new CSS file once you update the DevExtreme package to the new version.

**Comment by lifeinchords** [comment]
Hi @BingoRUS 

Thanks for the detailed reply and for the links to the .less files, that's a big help to understanding how it all works.

I see us using the 2nd option:`export-theme-meta` to `build-theme` so we can keep our application styles separate from 3rd party pieces. 

I still see a few issues:

- This CLI option is def better than no customization, though requires us to write a separate build step that runs an external exe, the DevEx CLI, on every compile. This adds time to the live reloading / dev server pipeline where every added step matters millisecond-wise. For production builds, I'm not sure how we'll handle errors, or if we'll run into other issues, but the idea is - it's another potential point of failure. The DevEx CLI itself feels like a build system, and overlaps in functionality/purpose with what a dedicated build system like Webpack is designed for. Am I missing anything - why build it in the first place, vs opening up the LESS files for customization?

- with this system the built CSS file includes the entire suite of DevEx components/widgets. We're using the Light theme now, it's ~450kb. That's a lot of extra unused file size.

- I noticed some  [global style declaration](https://github.com/DevExpress/DevExtreme/blob/18_2/styles/framework/common/framework.less). Is this being used as part of the CLI output?

---

I think CSS modules is forthcoming, but even opening up the LESS publicly would let DevEx users simply include the parts needed.

**Comment by BingoRUS** [comment]
Hi @lifeinchords,

We recommend ThemeBuilder to create your own color schemes for 2 DevExtreme themes: Generic and Material. It was available only as an online tool, so we decided to provide a way to include it to an automated build process. You don't have to call it on every build though, it's necessary to rebuild your color scheme stylesheet once you updated the DevExtreme library itself to a newer version. And we don't recommend doing it on the fly. So I don't see any point of failure in production here.
Why don't we provide LESS files for color scheme customization? It will be quite hard for you to maintain LESS files you changed when you update DevExtreme from one version to another. And it will be hard for our support team to help you with possible issues if you have a changed file.
We aware there are some reasons why people want us to distribute separate LESS files (you pointed to a size of the resulting stylesheet file), but right now we have some technical limitations: for example, [dataGrid.generic.less](https://github.com/DevExpress/DevExtreme/blob/18_2/styles/widgets/generic/dataGrid.generic.less) contains grid styles but doesn't contain styles for buttons in the toolbar and textboxes in case of editing. We need to rethink our styling system first to provide it in a truly modular way. Right now, we are collecting all feedback to plan future releases and we are very thankful for your comments.

**Comment by ezriharmusial** [comment]
Hi @BingoRUS,

True Modularity would be greatly Appreciated. As the Material Theme Light Blue based Themebuilder CSS file is a whopping 500k.

We already have customized our main Material Theme in SCSS outside DevExtreme. How do you advise we extract and integrate the DevExtreme SCSS for the Datagrid (including buttons, toolbar, textboxes) only? Is this possible with the DevExtreme CLI tool?

**Comment by Anjanas-dh** [comment]
When I generate a custom theme css with the themebuilder, there is a reference to the icons/ folder of devextreme within a @font-face style:
`url(icons/dxicons.woff2)`

But because my generated theme css file is not in the node_modules folder of dexextreme, I get an error when compiling:
`\themes\dx.theme.css:1489:95: Can't resolve 'icons/dxicons.woff' in '\themes'`

Do you know how I can resolve this error? I'm using the custom theme in an Angular 8 project.

**Comment by BingoRUS** [comment]
Hi @Anjanas-dh,

You can use the `assetsBasePath` field in an input file with ThemeBuilder metadata to provide the full path to a folder that contains icons. Here is the example of how we use this option in our DevExtreme Angular Template: https://github.com/DevExpress/devextreme-angular-template/blob/master/src/themes/metadata.base.json#L4.

**Comment by BingoRUS** [comment]
@ezriharmusial with v19.2 release you will be able to use `--widgets` parameter to specify the only components you want to have in the result CSS file. Please check [this issue](https://github.com/DevExpress/DevExtreme/issues/9720) for more details.

**Comment by BingoRUS** [comment]
Thank you to everybody who gave us feedback on this feature. DevExtreme CLI commands for theme generation are available since the v18.2 release. I'm closing this thread. In the case of bugs or questions, feel free to create new GitHub issues or tickets in our Support Center. Also, please check the [Custom Style Bundles](https://github.com/DevExpress/DevExtreme/issues/9720) discussion to get more information about the planned CSS modules feature in DevExtreme.
